code/go/0chain.net/blobbercore/allocation/copyfilechange.go

```diff
@@ -102,52 +111,36 @@ func (rf *CopyFileChange) ApplyChange(ctx context.Context, change *AllocationCha
 
 func (rf *CopyFileChange) processCopyRefs(
 	ctx context.Context, srcRef, destRef *reference.Ref,
-	allocationRoot string, ts common.Timestamp) (fileRefs []*reference.Ref) {
-
-	if srcRef.Type == reference.DIRECTORY {
-		newRef := reference.NewDirectoryRef()
-		newRef.AllocationID = rf.AllocationID
-		newRef.Path = filepath.Join(destRef.Path, srcRef.Name)
-		newRef.ParentPath = destRef.Path
-		newRef.Name = srcRef.Name
-		newRef.CreatedAt = ts
-		newRef.UpdatedAt = ts
-		newRef.HashToBeComputed = true
-		destRef.AddChild(newRef)
-
+	allocationRoot string, ts common.Timestamp, fileIDMeta map[string]string) (
+	fileRefs []*reference.Ref, err error) {
+
+	newRef := *srcRef
+	newRef.ID = 0
+	newRef.Path = filepath.Join(destRef.Path, srcRef.Name)
+	fileID, ok := fileIDMeta[newRef.Path]
+	if !ok || fileID == "" {
+		return nil, common.NewError("invalid_parameter",
+			fmt.Sprintf("file path %s has no entry in fileID meta", newRef.Path))
+	}
+	newRef.FileID = fileID
+	newRef.ParentPath = destRef.Path
+	newRef.CreatedAt = ts
+	newRef.UpdatedAt = ts
+	newRef.HashToBeComputed = true
+	destRef.AddChild(&newRef)
+	if newRef.Type == reference.DIRECTORY {
 		for _, childRef := range srcRef.Children {
-			fileRefs = append(fileRefs, rf.processCopyRefs(ctx, childRef, newRef, allocationRoot, ts)...)
+			fRefs, err := rf.processCopyRefs(ctx, childRef, &newRef, allocationRoot, ts, fileIDMeta)
```

Comment:
There is no depth control on this recursion because it is possible to have any  numbers of nested directories which will be limited by 50K limitation.
Controlling nested dirs is never made to be the part of protocol.